File: 03-march/28-find-the-string-with-lcp/solution.java
class Solution {
    public String findTheString(int[][] lcp) {
        int n = lcp.length;
        char[] word = new char[n];

        int charIdx = 0;
        for (int i = 0; i < n; i++) {
            if (word[i] != 0) continue;
            if (charIdx >= 26) return "";
            word[i] = (char) ('a' + charIdx++);
            for (int j = i + 1; j < n; j++) {
                if (lcp[i][j] > 0) {
                    if (word[j] != 0 && word[j] != word[i]) return "";
                    word[j] = word[i];
                }
            }
        }

        int[][] dp = new int[n][n];
        for (int i = n - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                if (word[i] == word[j]) {
                    dp[i][j] = (i + 1 < n && j + 1 < n ? dp[i + 1][j + 1] : 0) + 1;
                }
                if (dp[i][j] != lcp[i][j]) return "";
            }
        }

        return new String(word);
    }
}

File: 03-march/28-find-the-string-with-lcp/solution.py
class Solution:
    def findTheString(self, lcp: List[List[int]]) -> str:
        n = len(lcp)
        word = [''] * n

        char_idx = 0
        for i in range(n):
            if word[i]:
                continue
            if char_idx >= 26:
                return ""
            word[i] = chr(ord('a') + char_idx)
            char_idx += 1
            for j in range(i + 1, n):
                if lcp[i][j] > 0:
                    if word[j] and word[j] != word[i]:
                        return ""
                    word[j] = word[i]

        dp = [[0] * n for _ in range(n)]
        for i in range(n - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                if word[i] == word[j]:
                    dp[i][j] = (dp[i + 1][j + 1] if i + 1 < n and j + 1 < n else 0) + 1
                if dp[i][j] != lcp[i][j]:
                    return ""

        return ''.join(word)
